### PR TITLE
common library

### PR DESCRIPTION
**Pull request by lenala**
Common library for Services Explorer

### CODE REVIEW

**Comment by leotilli** [comment]
The entire content of the common library is still under com.microsoftopentechnologies.**intellij** package. If the purpose of the library is to be ported to Eclipse (and other IDEs),  we should remove the intellij portion of the namespace and replace it by something agnostic of the IDE, but representative of the product.

**Comment by leotilli** [comment]
All the NPE validation has been taken out from the model classes and the AzureSDKManager interface (and implementations). It would be great if we could add that functionality back, maybe using our own @NotNull and @Nullable implementations, so we don't depend on IntelliJ specific ones.

**Comment by leotilli** [comment]
Office365 model classes and helpers remain specific to IntelliJ. Although this is not currently being used by the Service Explorer functionality, we should still move Office365 related code to the common class, since it could be used in the future.
This change shouldn't be defined by IDEs specific requirements, but take all available functionalities and reorganize them into common code and IDE specific code.

**Comment by leotilli** [comment]
In order to be able to load the project in IntelliJ and build it, I've had to edit both msopentech-tools-for-intellij.iml and common.iml to add missing dependencies.
msopentech-tools-for-intellij.iml required a reference to common.iml as a module dependency.
common.iml required library dependencies to com.nimbusds:nimbus-jose-jwt:2.11.0 and .cspack.

**Comment by leotilli** [comment]
The current status of the PR prevents it from being automatically merged.

**Comment by leotilli** [comment]
We are unable to list or create Custom APIs for Mobile Services, through Service Explorer. Is showing the following exception:

java.lang.NoSuchMethodException: com.microsoftopentechnologies.intellij.serviceexplorer.azure.mobileservice.CustomAPINode.<init>(com.microsoftopentechnologies.intellij.serviceexplorer.Node, com.microsoftopentechnologies.intellij.model.ms.CustomAPI)
    at java.lang.Class.getConstructor0(Class.java:2715)
    at java.lang.Class.getConstructor(Class.java:1659)
    at com.microsoftopentechnologies.intellij.serviceexplorer.azure.mobileservice.MobileServiceNode.loadServiceNode(MobileServiceNode.java:139)
    at com.microsoftopentechnologies.intellij.serviceexplorer.azure.mobileservice.MobileServiceNode.refreshItems(MobileServiceNode.java:78)
    at com.microsoftopentechnologies.intellij.serviceexplorer.Node.refreshItems(Node.java:353)
    at com.microsoftopentechnologies.intellij.serviceexplorer.BackgroundLoader.run(BackgroundLoader.java:58)
    at com.intellij.openapi.progress.impl.CoreProgressManager$TaskRunnable.run(CoreProgressManager.java:563)
    at com.intellij.openapi.progress.impl.CoreProgressManager$2.run(CoreProgressManager.java:152)
    at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:452)
    at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:402)
    at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:54)
    at com.intellij.openapi.progress.impl.CoreProgressManager.runProcess(CoreProgressManager.java:137)
    at com.intellij.openapi.progress.impl.ProgressManagerImpl$1.run(ProgressManagerImpl.java:126)
    at com.intellij.openapi.application.impl.ApplicationImpl$8.run(ApplicationImpl.java:405)
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:439)
    at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:303)
    at java.util.concurrent.FutureTask.run(FutureTask.java:138)
    at java.util.concurrent.ThreadPoolExecutor$Worker.runTask(ThreadPoolExecutor.java:895)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:918)
    at java.lang.Thread.run(Thread.java:662)
    at org.jetbrains.ide.PooledThreadExecutor$1$1.run(PooledThreadExecutor.java:56)

**Comment by lenala** [comment]
I am not sure what com.microsoftopentechnologies.intellij.model.AzureType class is used for (I did not find any usages), leaving it as is in IntelliJ module. Should it be removed?
Also, I did not find any usages for Office365Application class.

**Comment by leotilli** [comment]
I've performed reference and text searches to both com.microsoftopentechnologies.intellij.model.AzureType and com.microsoftopentechnologies.intellij.model.Office365Application classes and they are not being used; I believe it should be alright to delete them.

File: src/common/com/microsoftopentechnologies/intellij/helpers/UIHelper.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
+package com.microsoftopentechnologies.intellij.helpers;
+
+import javax.swing.*;
+
+public interface UIHelper {
+    void packAndCenterJDialog(JDialog form);
```
Comment (leotilli) [inline src/common/com/microsoftopentechnologies/intellij/helpers/UIHelper.java]: There are 23 references to packAndCenterJDialog, all from IntelliJ specific classes. This method is specific to Swing dependent implementations (IntelliJ) and it shouldn't be exposed as part of this interface. I would leave it public in UIHelperImpl and call it directly from IntelliJ specific classes.

File: src/common/com/microsoftopentechnologies/intellij/helpers/UIHelper.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+    void showException(final String message,
+                                     final Throwable ex,
+                                     final String title,
+                                     final boolean appendEx,
+                                     final boolean suggestDetail);
+
+    public void showError(String message, String title);
+
+    ImageIcon loadIcon(String name);
```
Comment (leotilli) [inline src/common/com/microsoftopentechnologies/intellij/helpers/UIHelper.java]: The same happens here, there are 9 references to loadIcon, all from IntelliJ specific classes.

File: src/common/com/microsoftopentechnologies/intellij/helpers/UIHelper.java [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+    String readableFileSize(long size);
```
Comment (leotilli) [inline src/common/com/microsoftopentechnologies/intellij/helpers/UIHelper.java]: Idem packAndCenterJDialog and loadIcon. Only one reference from com.microsoftopentechnologies.intellij.helpers.storage.BlobExplorerFileEditor

**Comment by leotilli** [comment]
LGTM

**Pull request merged**